src/control/drpc/failure.go

```diff
@@ -0,0 +1,115 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// GOVERNMENT LICENSE RIGHTS-OPEN SOURCE SOFTWARE
+// The Government's rights to use, modify, reproduce, release, perform, display,
+// or disclose this software are subject to the terms of the Apache License as
+// provided in Contract No. 8F-30005.
+// Any reproduction of computer software, computer software documentation, or
+// portions thereof marked with this legend must also reproduce the markings.
+
+package drpc
+
+// Failure represents a dRPC protocol failure.
+type Failure struct {
+	statusCode Status
+	message    string
+}
+
+// Error provides a descriptive string associated with the failure.
+func (e Failure) Error() string {
+	return e.message
+}
+
+// GetStatus provides a dRPC status code associated with the failure.
+func (e Failure) GetStatus() Status {
+	return e.statusCode
+}
+
+// NewFailure returns a Failure with the given status and a corresponding message.
+func NewFailure(status Status) Failure {
+	message := statusToString(status)
+	return Failure{
+		message:    message,
+		statusCode: status,
+	}
+}
+
+func statusToString(status Status) string {
+	switch status {
+	case Status_UNKNOWN_MODULE:
+		return "unrecognized module ID"
+	case Status_UNKNOWN_METHOD:
+		return "unrecognized method ID"
+	case Status_FAILED_UNMARSHAL_CALL:
+		return "failed to unmarshal incoming call"
+	case Status_FAILED_UNMARSHAL_PAYLOAD:
+		return "failed to unmarshal method-specific payload"
+	case Status_FAILED_MARSHAL:
+		return "failed to marshal response payload"
+	case Status_SUCCESS:
+		fallthrough
+	case Status_SUBMITTED:
+		return ""
+	}
+
+	return "an unknown error occurred"
+}
+
+// UnknownModuleFailure creates a Failure for unknown dRPC module.
+func UnknownModuleFailure() Failure {
+	return NewFailure(Status_UNKNOWN_MODULE)
+}
+
+// UnknownMethodFailure creates a Failure for unknown dRPC method.
+func UnknownMethodFailure() Failure {
+	return NewFailure(Status_UNKNOWN_METHOD)
+}
+
+// UnmarshalingCallFailure creates a Failure for a failed attempt to unmarshal an incoming call.
+func UnmarshalingCallFailure() Failure {
+	return NewFailure(Status_FAILED_UNMARSHAL_CALL)
+}
+
+// UnmarshalingPayloadFailure creates a Failure for a failed attempt to unmarshal a call payload.
+func UnmarshalingPayloadFailure() Failure {
+	return NewFailure(Status_FAILED_UNMARSHAL_PAYLOAD)
+}
+
+// MarshalingFailure creates a Failure for a failed attempt at marshaling a response.
+func MarshalingFailure() Failure {
+	return NewFailure(Status_FAILED_MARSHAL)
+}
```

Comment:
My initial reaction to these was to wonder why you didn't just declare them as variables... It'd be better to make them consts, of course, but go doesn't support non-numeric consts. I guess doing it this way ensures that there are no shenanigans as a result of messing with the global variables?

Reply:
Yeah. Without the ability to make them consts, it seemed like a bad idea to make them global variables. Could go either way, I guess, but I had the NewFailure factory method already anyway...